File: app/src/test/java/com/quran/labs/androidquran/presenter/QuranImportPresenterTest.java
package com.quran.labs.androidquran.presenter;

import android.content.ContentResolver;
import android.content.Context;
import android.net.Uri;
import android.os.ParcelFileDescriptor;

import com.quran.labs.androidquran.model.bookmark.BookmarkImportExportModel;
import com.quran.labs.androidquran.model.bookmark.BookmarkModel;

import org.junit.Before;
import org.junit.Test;

import java.io.ByteArrayInputStream;
import java.io.FileNotFoundException;
import java.io.InputStream;
import java.util.List;

import io.reactivex.observers.TestObserver;
import okio.BufferedSource;

import static com.google.common.truth.Truth.assertThat;
import static org.mockito.Matchers.any;
import static org.mockito.Matchers.anyString;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

public class QuranImportPresenterTest {

  private Context appContext;
  private QuranImportPresenter presenter;

  @Before
  public void setup() {
    appContext = mock(Context.class);
    BookmarkImportExportModel model = mock(BookmarkImportExportModel.class);
    presenter = new QuranImportPresenter(appContext, model, mock(BookmarkModel.class));
  }

  @Test
  public void testParseExternalFile() throws FileNotFoundException {
    InputStream is = new ByteArrayInputStream(new byte[32]);
    ContentResolver resolver = mock(ContentResolver.class);
    when(resolver.openInputStream(any(Uri.class))).thenReturn(is);
    when(appContext.getContentResolver()).thenReturn(resolver);

    TestObserver<BufferedSource> observer = new TestObserver<>();
    presenter.parseExternalFile(Uri.EMPTY)
        .subscribe(observer);
    observer.awaitTerminalEvent();
    observer.assertValueCount(1);
    observer.assertNoErrors();
    observer.assertComplete();

    List<BufferedSource> events = observer.values();
    assertThat(events).hasSize(1);
    assertThat(events.get(0)).isNotNull();
  }

  @Test
  public void testParseExternalFileNullIs() throws FileNotFoundException {
    ContentResolver resolver = mock(ContentResolver.class);
    when(resolver.openInputStream(any(Uri.class))).thenReturn(null);
    when(appContext.getContentResolver()).thenReturn(resolver);

    TestObserver<BufferedSource> observer = new TestObserver<>();
    presenter.parseExternalFile(Uri.EMPTY)
        .subscribe(observer);
    observer.awaitTerminalEvent();
    observer.assertValueCount(0);
    observer.assertNoErrors();
    observer.assertComplete();
  }

  @Test
  public void testParseUriNullFd() throws FileNotFoundException {
    ContentResolver resolver = mock(ContentResolver.class);
    when(resolver.openFileDescriptor(any(Uri.class), anyString())).thenReturn(null);
    when(appContext.getContentResolver()).thenReturn(resolver);

    TestObserver<BufferedSource> observer = new TestObserver<>();
    presenter.parseUri(Uri.EMPTY)
        .subscribe(observer);
    observer.awaitTerminalEvent();
    observer.assertComplete();
    observer.assertValueCount(0);
    observer.assertNoErrors();
  }

  @Test
  public void testParseUriWithException() throws FileNotFoundException {
    ParcelFileDescriptor pfd = mock(ParcelFileDescriptor.class);
    when(pfd.getFd()).thenReturn(-1);

    ContentResolver resolver = mock(ContentResolver.class);
    when(resolver.openFileDescriptor(any(Uri.class), anyString())).thenReturn(pfd);
    when(appContext.getContentResolver()).thenReturn(resolver);

    TestObserver<BufferedSource> observer = new TestObserver<>();
    presenter.parseUri(Uri.EMPTY)
        .subscribe(observer);
    observer.awaitTerminalEvent();
    observer.assertError(NullPointerException.class);
    observer.assertValueCount(0);
  }
}

File: app/src/test/java/com/quran/labs/androidquran/presenter/QuranImportPresenterTest.kt
package com.quran.labs.androidquran.presenter

import android.content.Context
import android.net.Uri
import androidx.test.core.app.ApplicationProvider
import com.google.common.truth.Truth.assertThat
import com.quran.data.dao.Settings
import com.quran.labs.androidquran.base.TestApplication
import com.quran.labs.androidquran.fakes.FakeBookmarksDao
import com.quran.labs.androidquran.fakes.FakeContentResolverOps
import com.quran.labs.androidquran.fakes.FakePageProvider
import com.quran.labs.androidquran.fakes.FakeReadingBookmarksDao
import com.quran.labs.androidquran.fakes.FakeRecentPagesDao
import com.quran.labs.androidquran.model.bookmark.BookmarkImportExportModel
import com.quran.labs.androidquran.model.bookmark.BookmarkJsonModel
import com.quran.labs.awaitTerminalEvent
import com.quran.mobile.bookmark.importdata.BookmarkBackupImportNormalizer
import com.quran.mobile.bookmark.importdata.MobileSyncImportData
import com.quran.mobile.bookmark.importdata.MobileSyncImportResult
import com.quran.mobile.bookmark.importdata.MobileSyncImporter
import com.quran.mobile.bookmark.model.ReadingBookmarkPageMapper
import com.quran.mobile.bookmark.time.DefaultMobileSyncTimestampProvider
import io.reactivex.rxjava3.observers.TestObserver
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableSharedFlow
import okio.BufferedSource
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner
import org.robolectric.annotation.Config
import java.io.ByteArrayInputStream
import java.io.FileNotFoundException
import java.io.InputStream

@RunWith(RobolectricTestRunner::class)
@Config(application = TestApplication::class, sdk = [33])
class QuranImportPresenterTest {

  private lateinit var context: Context
  private lateinit var importExportModel: BookmarkImportExportModel

  @Before
  fun setup() {
    context = ApplicationProvider.getApplicationContext()
    val settings = FakeSettings()
    val pageMapper = ReadingBookmarkPageMapper(
      settings = settings,
      pageProviders = mapOf("madani" to FakePageProvider()),
      fallbackPageType = "madani"
    )
    importExportModel = BookmarkImportExportModel(
      context,
      BookmarkJsonModel(),
      FakeBookmarksDao(),
      FakeRecentPagesDao(),
      FakeReadingBookmarksDao(),
      settings,
      BookmarkBackupImportNormalizer(
        context,
        pageMapper,
        DefaultMobileSyncTimestampProvider()
      ),
      FakeMobileSyncImporter()
    )
  }

  // ---- parseExternalFile tests ----

  @Test
  @Throws(FileNotFoundException::class)
  fun testParseExternalFile() {
    val uri = Uri.parse("content://quran.test/backup")
    val stream: InputStream = ByteArrayInputStream(ByteArray(32))
    val fakeOps = FakeContentResolverOps(inputStream = stream)

    val presenter = QuranImportPresenter(context, importExportModel, fakeOps)

    val observer = TestObserver<BufferedSource>()
    presenter.parseExternalFile(uri).subscribe(observer)
    observer.awaitTerminalEvent()
    observer.assertValueCount(1)
    observer.assertNoErrors()
    observer.assertComplete()

    val events = observer.values()
    assertThat(events).hasSize(1)
    assertThat(events[0]).isNotNull()
  }

  @Test
  @Throws(FileNotFoundException::class)
  fun testParseExternalFileNullIs() {
    val uri = Uri.parse("content://quran.test/backup")
    // FakeContentResolverOps returns null inputStream by default
    val fakeOps = FakeContentResolverOps()

    val presenter = QuranImportPresenter(context, importExportModel, fakeOps)

    val observer = TestObserver<BufferedSource>()
    presenter.parseExternalFile(uri).subscribe(observer)
    observer.awaitTerminalEvent()
    observer.assertValueCount(0)
    observer.assertNoErrors()
    observer.assertComplete()
  }

  // ---- parseUri tests ----

  @Test
  @Throws(FileNotFoundException::class)
  fun testParseUriNullFd() {
    val uri = Uri.parse("content://quran.test/backup")
    // FakeContentResolverOps returns null fileDescriptor by default
    val fakeOps = FakeContentResolverOps()

    val presenter = QuranImportPresenter(context, importExportModel, fakeOps)

    val observer = TestObserver<BufferedSource>()
    presenter.parseUri(uri).subscribe(observer)
    observer.awaitTerminalEvent()
    observer.assertComplete()
    observer.assertValueCount(0)
    observer.assertNoErrors()
  }

  @Test
  @Throws(FileNotFoundException::class)
  fun testParseUriWithException() {
    val uri = Uri.parse("content://quran.test/backup")
    val fakeOps = FakeContentResolverOps(fileDescriptorException = NullPointerException())

    val presenter = QuranImportPresenter(context, importExportModel, fakeOps)

    val observer = TestObserver<BufferedSource>()
    presenter.parseUri(uri).subscribe(observer)
    observer.awaitTerminalEvent()
    observer.assertError(NullPointerException::class.java)
    observer.assertValueCount(0)
  }

  private class FakeMobileSyncImporter : MobileSyncImporter {
    override suspend fun importData(
      data: MobileSyncImportData,
      deleteExisting: Boolean
    ): MobileSyncImportResult {
      return MobileSyncImportResult(
        bookmarksImported = data.bookmarks.size,
        collectionsImported = data.collections.size,
        collectionBookmarksImported = data.collectionBookmarks.size,
        readingSessionsImported = data.readingSessions.size,
        readingBookmarkImported = data.readingBookmark != null
      )
    }
  }

  private class FakeSettings : Settings {
    private val preferences = MutableSharedFlow<String>(extraBufferCapacity = 1)
    private var pageType = "madani"

    override suspend fun setVersion(version: Int) = Unit

    override suspend fun setShouldOverlayPageInfo(shouldOverlay: Boolean) = Unit

    override suspend fun lastPage(): Int = 1

    override suspend fun isNightMode(): Boolean = false

    override suspend fun nightModeTextBrightness(): Int = 0

    override suspend fun nightModeBackgroundBrightness(): Int = 0

    override suspend fun shouldShowHeaderFooter(): Boolean = false

    override suspend fun shouldShowBookmarks(): Boolean = false

    override suspend fun pageType(): String = pageType

    override suspend fun setPageType(pageType: String) {
      this.pageType = pageType
      preferences.emit("pageType")
    }

    override suspend fun showSidelines(): Boolean = false

    override suspend fun setShowSidelines(show: Boolean) = Unit

    override suspend fun showLineDividers(): Boolean = false

    override suspend fun setShouldShowLineDividers(show: Boolean) = Unit

    override suspend fun setAyahTextSize(value: Int) = Unit

    override suspend fun translationTextSize(): Int = 0

    override fun preferencesFlow(): Flow<String> = preferences
  }
}
